👩‍🌾 test_tutorial_parameter_events_async__rmw_connext_cpp regression on Windows Release

This test has failed in the last 5/6 most recent builds (since 2020-10-28), and it was passing consistently before that:

https://ci.ros2.org/view/nightly/job/nightly_win_rel/1740/testReport/(root)/projectroot/test_tutorial_parameter_events_async__rmw_connext_cpp/history/

<details><summary>Test failure output</summary>
<p>

```
-- run_test.py: extra environment variables:
 - RCL_ASSERT_RMW_ID_MATCHES=rmw_connext_cpp
 - RMW_IMPLEMENTATION=rmw_connext_cpp
-- run_test.py: invoking following command in 'C:\ci\ws\build\demo_nodes_cpp':
 - C:/Python38/python.exe -m launch_testing.launch_test C:/ci/ws/build/demo_nodes_cpp/test_parameter_events_async__rmw_connext_cpp_Release.py --junit-xml=C:/ci/ws/build/demo_nodes_cpp/test_results/demo_nodes_cpp/test_tutorial_parameter_events_async__rmw_connext_cpp.xunit.xml --package-name=demo_nodes_cpp
[INFO] [launch]: All log files can be found below C:\Users\ContainerAdministrator\.ros\log\2020-11-02-11-25-42-031390-c18fd2498ae4-5020
[INFO] [launch]: Default logging verbosity is set to INFO
test_processes_output (demo_nodes_cpp.TestExecutablesTutorial)
Test all processes output against expectations. ... [INFO] [test_executable_0-1]: process started with pid [7068]
[INFO] [python.exe-2]: process started with pid [16352]
[test_executable_0-1] COMMENDBeWriterService_assertRemoteReader:Discovered remote reader with GUID 0X10F777D,0X14380000,0X1000000,0X200C7 using a non-addressable locator.
[test_executable_0-1] This can occur if the transport is not installed and/or enabled in the local participant.
[test_executable_0-1] See https://community.rti.com/kb/what-does-cant-reach-locator-error-message-mean for additional info.
[test_executable_0-1] can't reach: locator: 16://5577:7D00:0000:0000:0000:0000:0000:0000:7662
[test_executable_0-1] COMMENDBeWriterService_assertRemoteReader:Discovered remote reader with GUID 0X10F777D,0XB8260000,0X1000000,0X200C7 using a non-addressable locator.
[test_executable_0-1] This can occur if the transport is not installed and/or enabled in the local participant.
[test_executable_0-1] See https://community.rti.com/kb/what-does-cant-reach-locator-error-message-mean for additional info.
[test_executable_0-1] can't reach: locator: 16://5577:7D00:0000:0000:0000:0000:0000:0000:7664
[test_executable_0-1] COMMENDSrWriterService_assertRemoteReader:Discovered remote reader with GUID 0X10F777D,0X14380000,0X1000000,0X3C7 using a non-addressable locator.
[test_executable_0-1] This can occur if the transport is not installed and/or enabled in the local participant.
[test_executable_0-1] See https://community.rti.com/kb/what-does-cant-reach-locator-error-message-mean for additional info.
[test_executable_0-1] can't reach: locator: 16://5577:7D00:0000:0000:0000:0000:0000:0000:7662
[test_executable_0-1] COMMENDSrWriterService_assertRemoteReader:Discovered remote reader with GUID 0X10F777D,0XB8260000,0X1000000,0X3C7 using a non-addressable locator.
[test_executable_0-1] This can occur if the transport is not installed and/or enabled in the local participant.
[test_executable_0-1] See https://community.rti.com/kb/what-does-cant-reach-locator-error-message-mean for additional info.
[test_executable_0-1] can't reach: locator: 16://5577:7D00:0000:0000:0000:0000:0000:0000:7664
[test_executable_0-1] COMMENDSrWriterService_assertRemoteReader:Discovered remote reader with GUID 0X10F777D,0X14380000,0X1000000,0X4C7 using a non-addressable locator.
[test_executable_0-1] This can occur if the transport is not installed and/or enabled in the local participant.
[test_executable_0-1] See https://community.rti.com/kb/what-does-cant-reach-locator-error-message-mean for additional info.
[test_executable_0-1] can't reach: locator: 16://5577:7D00:0000:0000:0000:0000:0000:0000:7662
[test_executable_0-1] COMMENDSrWriterService_assertRemoteReader:Discovered remote reader with GUID 0X10F777D,0XB8260000,0X1000000,0X4C7 using a non-addressable locator.
[test_executable_0-1] This can occur if the transport is not installed and/or enabled in the local participant.
[test_executable_0-1] See https://community.rti.com/kb/what-does-cant-reach-locator-error-message-mean for additional info.
[test_executable_0-1] can't reach: locator: 16://5577:7D00:0000:0000:0000:0000:0000:0000:7664
[test_executable_0-1] 
[test_executable_0-1] Parameter event:
[test_executable_0-1]  new parameters:
[test_executable_0-1]  changed parameters:
[test_executable_0-1]   foo
[test_executable_0-1]  deleted parameters:
[test_executable_0-1] 
[test_executable_0-1] 
[test_executable_0-1] Parameter event:
[test_executable_0-1]  new parameters:
[test_executable_0-1]  changed parameters:
[test_executable_0-1]   bar
[test_executable_0-1]  deleted parameters:
[test_executable_0-1] 
[test_executable_0-1] 
[test_executable_0-1] Parameter event:
[test_executable_0-1]  new parameters:
[test_executable_0-1]  changed parameters:
[test_executable_0-1]   baz
[test_executable_0-1]  deleted parameters:
[test_executable_0-1] 
[test_executable_0-1] 
[test_executable_0-1] Parameter event:
[test_executable_0-1]  new parameters:
[test_executable_0-1]  changed parameters:
[test_executable_0-1]   foobar
[test_executable_0-1]  deleted parameters:
[test_executable_0-1] 
[test_executable_0-1] 
[test_executable_0-1] Parameter event:
[test_executable_0-1]  new parameters:
[test_executable_0-1]  changed parameters:
[test_executable_0-1]   foo
[test_executable_0-1]  deleted parameters:
[test_executable_0-1] 
[test_executable_0-1] 
[test_executable_0-1] Parameter event:
[test_executable_0-1]  new parameters:
[test_executable_0-1]  changed parameters:
[test_executable_0-1]   bar
[test_executable_0-1]  deleted parameters:
[test_executable_0-1] 
[INFO] [test_executable_0-1]: process has finished cleanly [pid 7068]
[WARNING] [python.exe-2]: 'SIGINT' sent to process[python.exe-2] not supported on Windows, escalating to 'SIGTERM'
[INFO] [python.exe-2]: sending signal 'SIGTERM' to process[python.exe-2]
[ERROR] [python.exe-2]: process has died [pid 16352, exit code 1, cmd 'C:\Python38\python.exe -c 
try:
    while True:
        pass
except KeyboardInterrupt:
    pass
'].
FAIL

======================================================================
FAIL: test_processes_output (demo_nodes_cpp.TestExecutablesTutorial)
Test all processes output against expectations.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\ci\ws\build\demo_nodes_cpp\test_parameter_events_async__rmw_connext_cpp_Release.py", line 55, in test_processes_output
    proc_output.assertWaitFor(
  File "C:\ci\ws\install\Lib\site-packages\launch_testing\io_handler.py", line 146, in assertWaitFor
    assert success, 'Waiting for output timed out'
AssertionError: Waiting for output timed out

```

</p>
</details>

--- a/demo_nodes_cpp/src/parameters/parameter_events_async.cpp
+++ b/demo_nodes_cpp/src/parameters/parameter_events_async.cpp
@@ -63,6 +63,18 @@ public:
     // Setup callback for changes to parameters.
     parameter_event_sub_ = parameters_client_->on_parameter_event(on_parameter_event_callback);
 
+    // Even though this is in the same node, we still have to wait for the
+    // service to be available before declaring parameters (otherwise there is
+    // a chance we'll miss some events in the callback).
+    while (!parameters_client_->wait_for_service(1s)) {
+      if (!rclcpp::ok()) {
+        RCLCPP_ERROR(this->get_logger(), "interrupted while waiting for the service. exiting.");
+        rclcpp::shutdown();
+        return;
+      }
+      RCLCPP_INFO(this->get_logger(), "service not available, waiting again...");
+    }
+
     // Declare parameters that may be set on this node
     this->declare_parameter("foo");
     this->declare_parameter("bar");

